Objects/rangeobject.c

```diff
@@ -970,7 +971,18 @@ longrangeiter_setstate(longrangeiterobject *r, PyObject *state)
         if (cmp > 0)
             state = r->len;
     }
-    Py_XSETREF(r->index, Py_NewRef(state));
+    PyObject *new_len = PyNumber_Subtract(r->len, state);
+    if (new_len == NULL)
+        return NULL;
+    Py_SETREF(r->len, new_len);
```

Comment:
This is a little scary -- if the new length is set but computing the new start fails, and the original iterator is still accessible, it will be truncated at the end rather than at the start. Let's update both fields after all errors have been checked.

Ditto in next().

Reply:
Seems that it never happens in normal code. You have to call `__setstate__()` manually with the out-of-range value.